contrib/babelfishpg_tsql/src/pl_exec-2.c

```diff
@@ -3287,3 +3288,98 @@ get_insert_bulk_kilobytes_per_batch()
+	ListCell   *parsetree_item;
+	char	   *dbname = get_cur_db_name();
+	char	   *login = GetUserNameFromId(GetSessionUserId(), false);
+	bool		login_is_db_owner;
+	Oid			datdba;
+	char		*rolname;
+	char		*schema_name;
+	ListCell   *lc;
+	ListCell	*lc1;
+	Oid 		schemaOid;
+
+	/*
+	 * If the login is not the db owner or the login is not the member of
+	 * sysadmin or login is not the schema owner, then it doesn't have the permission to GRANT/REVOKE.
+	 */
+	login_is_db_owner = 0 == strncmp(login, get_owner_of_db(dbname), NAMEDATALEN);
+	datdba = get_role_oid("sysadmin", false);
+	schema_name = get_physical_schema_name(dbname, stmt->schema_name);
+	schemaOid = LookupExplicitNamespace(schema_name, true);
+
+	if (!OidIsValid(schemaOid))
+			ereport(ERROR,
+					(errcode(ERRCODE_UNDEFINED_SCHEMA),
+					 errmsg("schema \"%s\" does not exist",
+							schema_name)));
+	
+	if (!is_member_of_role(GetSessionUserId(), datdba) && !login_is_db_owner && !pg_namespace_ownercheck(schemaOid, GetUserId()))
+		ereport(ERROR,
+				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+				 errmsg("Cannot find the schema \"%s\", because it does not exist or you do not have permission.", stmt->schema_name)));
+
+	foreach(lc1, stmt->privileges)
+	{
+		char	*priv_name = (char *) lfirst(lc1);
+		foreach(lc, stmt->grantees)
+		{
+			char	   *grantee_name = (char *) lfirst(lc);
+			Oid	role_oid;
+			bool		grantee_is_db_owner;
+			rolname	= get_physical_user_name(dbname, grantee_name);
+			role_oid = get_role_oid(rolname, true);
+			grantee_is_db_owner = 0 == strncmp(grantee_name, get_owner_of_db(dbname), NAMEDATALEN);
+	
+			if (pg_namespace_ownercheck(schemaOid, role_oid) || is_member_of_role(role_oid, datdba) || grantee_is_db_owner)
+				ereport(ERROR,
+					(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+					 errmsg("Cannot grant, deny, or revoke permissions to sa, dbo, entity owner, information_schema, sys, or yourself.")));
+
+			parsetree_list = gen_grantschema_subcmds(schema_name, rolname, stmt->is_grant, stmt->with_grant_option, priv_name);
+			/* Run all subcommands */
+			foreach(parsetree_item, parsetree_list)
+			{
+				Node	   *stmt = ((RawStmt *) lfirst(parsetree_item))->stmt;
+				PlannedStmt *wrapper;
+
+				/* need to make a wrapper PlannedStmt */
+				wrapper = makeNode(PlannedStmt);
+				wrapper->commandType = CMD_UTILITY;
+				wrapper->canSetTag = false;
+				wrapper->utilityStmt = stmt;
+				wrapper->stmt_location = 0;
+				wrapper->stmt_len = 0;
+
+				/* do this step */
+				ProcessUtility(wrapper,
+							"(GRANT SCHEMA )",
+							false,
+							PROCESS_UTILITY_SUBCOMMAND,
+							NULL,
+							NULL,
+							None_Receiver,
+							NULL);
+
+				/* make sure later steps can see the object created here */
+				CommandCounterIncrement();
+			}
+			/* Add entry for each grant statement. */
+			if (stmt->is_grant && !check_bbf_schema_for_entry(stmt->schema_name, "ALL", priv_name, rolname))
+				add_entry_to_bbf_schema(stmt->schema_name, "ALL", priv_name, rolname, NULL);
```

Comment:
Why is permission and grantee column of table not an array?  This is going to bloat our memory drastically.
In one query it is possible for the customer to assign m permissions to n grantees. In that case you are adding m*n rows.
In PG CATALOG we have these as an array with a bit of performance overhead